File: src/ir/print.ts
import * as ir from "./node";
import { TypeKind, ValueKind, InstrKind } from "./kind";

function typeStr(t: ir.Type): string {
  switch (t.kind) {
    case TypeKind.Int: return t.width ? `int${t.width}` : "int";
    case TypeKind.Float: return t.width ? `float${t.width}` : "float";
    case TypeKind.Bool: return "bool";
    case TypeKind.Ptr: return `ptr<${typeStr(t.elem!)}>`;
    case TypeKind.Array: return `[${t.width}]${typeStr(t.elem!)}`;
    case TypeKind.Struct: return t.name || "struct";
    case TypeKind.Void: return "void";
    case TypeKind.Error: return "error";
  }
}

function valueStr(v: ir.Value): string {
  switch (v.kind) {
    case ValueKind.Reg: return `%${v.name}`;
    case ValueKind.Const: return v.name;
    case ValueKind.Global: return `@${v.name}`;
  }
}

export function print(m: ir.Module): string {
  let out = `module ${m.name} {\n\n`;
  for (const ext of m.externs) {
    const params = ext.params.map(typeStr).join(", ");
    out += `  extern "${ext.abi}" fn @${ext.name}(${params}) -> ${typeStr(ext.ret)}\n`;
  }
  if (m.externs.length > 0) out += "\n";
  for (const f of m.funcs) {
    const params = f.params.map(p => `${valueStr(p)}: ${typeStr(p.type)}`).join(", ");
    const tparams = f.typeParams ? `[${f.typeParams.join(", ")}]` : "";
    out += `fn @${f.name}${tparams}(${params}) -> ${typeStr(f.ret)} {\n`;
    for (const b of f.blocks) {
      out += `\n${b.label}:\n`;
      for (const i of b.instrs) {
        out += `  ${printInstr(i)}\n`;
      }
      out += `  ${printInstr(b.term)}\n`;
    }
    out += "}\n\n";
  }
  return out.trim() + "\n}";
}

function printInstr(i: ir.Instruction): string {
  const dest = i.dest ? `${valueStr(i.dest)} = ` : "";
  switch (i.kind) {
    case InstrKind.Alloc:
      return `${dest}alloc ${i.region} ${typeStr(i.dest!.type)}`;
    case InstrKind.Load:
      return `${dest}load ${valueStr(i.ptr!)}`;
    case InstrKind.Store:
      return `store ${valueStr(i.val!)}, ${valueStr(i.ptr!)}`;
    case InstrKind.Add:
      return `${dest}add ${valueStr(i.val!)}, ${valueStr(i.val2!)}`;
    case InstrKind.Sub:
      return `${dest}sub ${valueStr(i.val!)}, ${valueStr(i.val2!)}`;
    case InstrKind.Mul:
      return `${dest}mul ${valueStr(i.val!)}, ${valueStr(i.val2!)}`;
    case InstrKind.Div:
      return `${dest}div ${valueStr(i.val!)}, ${valueStr(i.val2!)}`;
    case InstrKind.Call: {
      const args = (i.args || []).map(valueStr).join(", ");
      const targs = i.typeArgs ? `[${i.typeArgs.map(typeStr).join(", ")}]` : "";
      return `${dest}call @${i.func}${targs}(${args})`;
    }
    case InstrKind.Jump:
      return `jump ${i.label}`;
    case InstrKind.Cjump:
      return `cjump ${valueStr(i.cond!)}, ${i.trueLabel}, ${i.falseLabel}`;
    case InstrKind.Return:
      return i.value ? `return ${valueStr(i.value)}` : "return";
    case InstrKind.Panic:
      return `panic "${i.msg}"`;
    case InstrKind.Try:
      return `${dest}try ${valueStr(i.expr!)}, ${i.handler}`;
    case InstrKind.Phi:
      const sources = (i.blocks || []).map(b => `[${b.label}: ${valueStr(b.value)}]`).join(", ");
      return `${dest}phi ${sources}`;
    case InstrKind.Eval:
      return `${dest}eval "${i.evalBody}"`;
    case InstrKind.Reflect:
      return `${dest}reflect ${typeStr(i.typeArg!)}`;
    case InstrKind.Embed:
      return `${dest}embed "${i.path}"`;
  }
}

File: test/ir/print.nim
import std/[json, strutils, tables]
import ../../src/ir/[kind, node, print]

let integer = `Type`(kind: TypeKind.Int, width: 64)
let register = Value(kind: ValueKind.Reg, name: "1", `type`: integer)
let constant = Value(kind: ValueKind.Const, name: "41", `type`: integer)
let add = Instruction(kind: InstrKind.Add, dest: register, val: constant,
  val2: Value(kind: ValueKind.Const, name: "1", `type`: integer))
let finish = Instruction(kind: InstrKind.Return, value: register)
let function = Function(name: "main", params: @[], ret: integer,
  blocks: @[Block(label: "entry", instrs: @[add], term: finish)])

let legacy = Module(name: "sample", funcs: @[function])
let listing = print(legacy)
doAssert listing.contains("module sample")
doAssert listing.contains("%1 = add 41, 1")
doAssert listing.contains("return %1")

let interchange = Module(name: "sample", version: 1, stage: "@foo",
  funcs: @[function])
let encoded = encode(interchange).parseJson
doAssert encoded["format"].getStr == "foo.ir"
doAssert encoded["version"].getInt == 1
doAssert encoded["functions"][0]["name"].getStr == "main"
doAssert encoded["constants"].len == 2

let memory = `Type`(kind: TypeKind.Memory)
let effect = Instruction(kind: InstrKind.Call, `func`: "write",
  args: @[constant], effects: @["write"], memory: Memory(input: "$memory0", output: "$memory1"),
  span: SourceSpan(file: "main.iv", start: 3, `end`: 9))
let rich = Module(name: "sample", version: 1, stage: "@foo",
  unitPackage: "pkg", unitPath: "src/main.iv", requires: "system",
  target: {"os": "linux", "arch": "x86_64"}.toTable,
  storage: @[Storage(name: "counter", public: true, value: constant)],
  regions: @[Region(id: "main.scope", kind: "scope")],
  traces: @[Trace(id: "main.trace", `function`: "main")],
  native: @[NativeContract(id: "native0", stage: "@c", code: "x", abi: "foo:1", effects: @["unknown"])],
  residue: @[Residue(id: "residue0", operation: "eval", state: "pending", payload: "x")],
  externs: @[Extern(name: "write", params: @[integer], ret: `Type`(kind: TypeKind.Void), abi: "c")],
  funcs: @[Function(name: "main", params: @[], ret: `Type`(kind: TypeKind.Void),
    blocks: @[Block(label: "entry", params: @[Value(kind: ValueKind.Reg, name: "$memory0", `type`: memory)],
      instrs: @[effect], term: Instruction(kind: InstrKind.Return, memory: Memory(input: "$memory1")))])])
let complete = encode(rich).parseJson
doAssert complete["unit"]["package"].getStr == "pkg"
doAssert complete["target"]["os"].getStr == "linux"
doAssert complete["storage"].len == 1
doAssert complete["regions"].len == 1
doAssert complete["traces"].len == 1
doAssert complete["native"].len == 1
doAssert complete["residue"].len == 1
let encodedEffect = complete["functions"][0]["blocks"][0]["instructions"][0]["attributes"]
doAssert encodedEffect["effects"][0].getStr == "write"
doAssert encodedEffect["memory"]["input"].getStr == "$memory0"
doAssert encodedEffect["span"]["file"].getStr == "main.iv"

echo "IR print parity: ok"
